genesis/engine/solvers/rigid/gjk_decomp.py

```diff
@@ -2914,9 +2973,839 @@ def support_driver(self, direction, i_g, i_b, i_o, shrink_sphere):
+            # vertex positions in a different way than the general mesh.
+            v = ti.Vector(
+                [
+                    (1.0 if (i_v & 1 == 1) else -1.0) * self._solver.geoms_info[i_g].data[0] * 0.5,
+                    (1.0 if (i_v & 2 == 2) else -1.0) * self._solver.geoms_info[i_g].data[1] * 0.5,
+                    (1.0 if (i_v & 4 == 4) else -1.0) * self._solver.geoms_info[i_g].data[2] * 0.5,
+                ],
+                dt=gs.ti_float,
+            )
+        elif geom_type == gs.GEOM_TYPE.MESH:
+            vert_start = self._solver.geoms_info.vert_start[i_g]
+            v = self._solver.verts_info[vert_start + i_v].init_pos
+
+        # Transform the vertex position to the world frame
+        v = gu.ti_transform_by_trans_quat(v, geom_state.pos, geom_state.quat)
+
+        return v
+
+    @ti.func
+    def func_safe_gjk_triangle_info(self, i_b, i_ta, i_tb, i_tc, i_apex):
+        """
+        Compute normal and signed distance of the triangle face on the simplex from the origin.
+
+        The triangle is defined by the vertices [i_ta], [i_tb], and [i_tc], and the apex is used to orient the triangle
+        normal, so that it points outward from the simplex. Thus, if the origin is inside the simplex in terms of this
+        triangle, the signed distance will be positive.
+        """
+        vertex_1 = self.simplex_vertex[i_b, i_ta].mink
+        vertex_2 = self.simplex_vertex[i_b, i_tb].mink
+        vertex_3 = self.simplex_vertex[i_b, i_tc].mink
+        apex_vertex = self.simplex_vertex[i_b, i_apex].mink
+
+        # This normal is guaranteed to be non-zero because we build the simplex avoiding degenerate vertices.
+        normal = (vertex_3 - vertex_1).cross(vertex_2 - vertex_1).normalized()
+
+        # Reorient the normal to point outward from the simplex
+        if normal.dot(apex_vertex - vertex_1) > 0.0:
+            normal = -normal
+
+        # Compute the signed distance from the origin to the triangle plane
+        sdist = normal.dot(vertex_1)
+
+        return normal, sdist
+
+    @ti.func
+    def func_safe_gjk_support(self, i_ga, i_gb, i_b, dir):
+        """
+        Find support points on the two objects using [dir] to use in the [safe_gjk] algorithm.
+
+        This is a more robust version of the support function that finds only one pair of support points, because this
+        function perturbs the support direction to find the best support points that guarantee non-degenerate simplex
+        in the GJK algorithm.
+
+        Parameters:
+        ----------
+        dir: gs.ti_vec3
+            The unit direction in which to find the support points, from [ga] (obj 1) to [gb] (obj 2).
+        """
+        obj1 = gs.ti_vec3(0, 0, 0)
+        obj2 = gs.ti_vec3(0, 0, 0)
+        id1 = -1
+        id2 = -1
+        mink = obj1 - obj2
+
+        for i in range(9):
+            # px, py, pz = gs.ti_float(0.0), gs.ti_float(0.0), gs.ti_float(0.0)
```

Comment:
What is this? Remove if useless, or add a `# FIXME:` statement to explain why it is kept in the code.